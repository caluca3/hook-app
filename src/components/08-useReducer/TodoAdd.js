import React from 'react'
import { useForm } from '../../hooks/useForm';

//Para enviar un parametro creferenciado desde un padre se coloca entre llaves
//luego se define en el componente padre
//<componente 
// funcionenviada={funcionenviada}
// />
export const TodoAdd = ({handleADDTodo}) => {

    const [{description},handleInputChange,reset] = useForm({
        description:'',
    });

    const handleADD=(e)=>{
        e.preventDefault();
        
        if (description.trim().length <=1) {
            return;
        }
        const newTodo ={
            id:new Date().getTime(),
            desc:description,
            done:false
        }
       

        handleADDTodo(newTodo);
        reset();
    }

  return (
    <>
         <form 
            onSubmit={handleADD}className='d-grid gap-2'>
                <input
                className='form-control'
                type='text'
                placeholder='Agregar..'
                name='description'
                autoComplete='off'
                value={description}
                onChange={handleInputChange}/>
                <button 
                type='submit'
                className='btn btn-outline-primary' >
                    Agregar
                </button>
            </form>    
            
    </>
  )
}
