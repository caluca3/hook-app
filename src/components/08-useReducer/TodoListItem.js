import React from 'react'


export const TodoListItem = ({todo,i,handleDelete,handleToggle}) => {
    
    
return (<>
<li
    key= {todo.id}
    className='list-group-item'
 >
    <p
     onClick={()=>handleToggle(todo.id)}
     className={`${todo.done &&'complete'}`}
    > {i+1} {todo.desc}
    
    </p>

    <button className='btn btn-danger'
    /*Se envia la funcion con un parametro ya que asi fue creada */

       onClick={()=>handleDelete(todo.id)}>
       Borrar

    </button>
</li>
    
    
</>)
}
