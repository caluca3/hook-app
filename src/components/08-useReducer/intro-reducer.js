//Se envia un obejto dentro de un arreglo
const initialState =[{
    id:1,
    todo:'Comprar pan',
    done:false
}];

const todoReducer= (state=initialState,action)=>{
    if (action?.type==='agg') {
     return [...state, action.payload];
    }
    return state;
}

let todos=todoReducer();

//NUNCA se usa .push En REACT!!.a menos que sea con
//el useRef ya que el .push muta el objeto

const newTodo={
    id:2,
    todo:'comprar leche',
    done:false
}

const action={ 
    //siempre se usa de manera estandarizada
    //tanto el type(la accion que hace) como el payload
    type:'agg',
    payload: newTodo
}

todos=todoReducer(todos,action);

console.log( todos);