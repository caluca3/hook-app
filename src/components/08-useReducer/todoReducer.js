
//si el state no esta inicializado lo iniciamos como un 
//arreglo vacio

export const todoReducer =(state=[],action)=>{
    switch (action.type) {
       case 'add':
            return [...state, action.payload];
        
        case 'delete':
            return state.filter(todo => todo.id !== action.payload);
    
//nueva manera usando el operador ternario 
//ya que siempre se necesitan dos ciondiciones
        case 'toggle':
            return state.map(todo=>
                (todo.id===action.payload)
                ?
                {...todo,done:!todo.done}
                :todo
                )




            //Asi se manejaba antes
        case 'toggle-old':
            return state.map(todo =>{

                if (todo.id===action.payload) {
                    return{
                        ...todo,
                        done:!todo.done
                    }
                    }else{
                        return todo;
                    }
                })
    
        default:
        return state;
    }
}