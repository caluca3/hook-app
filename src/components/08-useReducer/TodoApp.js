import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import './styles.css';


export const init = ()=>{ 
    //asi se realiza para no guardar en localstrogae
    //  return [{
    //      id:new Date().getTime(),
    //      desc:'Aprender React',
    //      done:false
    //  }];
    //si exiten los todo traelos y si no hay es null
    //si es null trae un arreglo vacio
    return JSON.parse(localStorage.getItem('todos')) || [];
    }

export const TodoApp= () => {
    
    const [todos,dispatch] = useReducer(todoReducer,[],init);
    



   useEffect(() => {
     localStorage.setItem('todos',JSON.stringify(todos));
   }, [todos]);
   
   
   const handleDelete =(todoId,)=>{
    const action ={
        type:'delete',
        payload:todoId
    }
    dispatch(action);
     }

     const handleToggle =(todoId)=>{
    //Se usa la funcion dispatch para enviar la accion al reducer en talcaso 
    //que esta use dispatch
    //ya que son pasados por referencia
        dispatch({
            type:'toggle',
            payload:todoId
        });
     }

const handleADDTodo =(newTodo)=>{
    
    dispatch({
        type:'add',
        payload:newTodo
    })
}
   

  return (<>
  
    <div>
      <h1>TodoApp ({todos.length})</h1>
    <hr/>

    <div className='row '>
        <div className='col-7'>

                <TodoList
                todos={todos}
                handleToggle={handleToggle}
                handleDelete={handleDelete}
                />

        </div>
    <div className='col-5 '>
            <h4>Agregar TODO</h4>
            <hr/>
           <TodoAdd
           handleADDTodo={handleADDTodo}/>
        </div>
    </div>
  </div>
  </>
  )
}
 