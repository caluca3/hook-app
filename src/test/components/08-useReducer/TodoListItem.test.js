import { act, renderHook }from "@testing-library/react-hooks";
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";
import { demoTodo } from '../../fixtures/demotodo';


const handleDelete=jest.fn();
const handleToggle=jest.fn();

describe('Pruebas en todolist', () => { 
    const {result} = renderHook(()=>TodoListItem({
        todo:demoTodo[0],
        i:0,
        handleDelete:{handleDelete},
        handleToggle:{handleToggle}
    }));


        test('should correctly', () => { 
        expect(result).toMatchSnapshot()
    });
    
    test('Debe llamar la funcion borrar', () => { 
        act(()=>{
            handleDelete(1);
        })
        expect(handleDelete).toHaveBeenCalledWith(demoTodo[0].id);
    });
    test('Debe llamar la funcion Toggle', () => { 
        act(()=>{
            handleToggle(1);
        })
        expect(handleToggle).toHaveBeenCalledWith(demoTodo[0].id);
    });
    test('Muestra el texto correctamente', () => {
        const {p} = result.current.props;
      
        const {parr} = result.all.values;
        expect(parr).toEqual(`1.${demoTodo[0].desc}`)
    })
    
})