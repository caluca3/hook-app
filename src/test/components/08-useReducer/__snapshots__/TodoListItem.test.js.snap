// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en todolist should correctly 1`] = `
Object {
  "all": Array [
    <React.Fragment>
      <li
        className="list-group-item"
      >
        <p
          className="false"
          onClick={[Function]}
        >
           
          1
           
          Aprender
        </p>
        <button
          className="btn btn-danger"
          onClick={[Function]}
        >
          Borrar
        </button>
      </li>
    </React.Fragment>,
  ],
  "current": <React.Fragment>
    <li
      className="list-group-item"
    >
      <p
        className="false"
        onClick={[Function]}
      >
         
        1
         
        Aprender
      </p>
      <button
        className="btn btn-danger"
        onClick={[Function]}
      >
        Borrar
      </button>
    </li>
  </React.Fragment>,
  "error": undefined,
}
`;
